File: Inventor-Training-Material-0c232f96f36e1e0bed31b07a6a6ebb901a31dd40/Module 21 - Installer/Registry_Free_AddIn_Demo_C#/Registry_Free_AddIn_Demo/AddinInstaller1.cs
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Copyright (c) Autodesk, Inc. All rights reserved 
// Written by Philippe Leefsma 2011 - ADN/Developer Technical Services
//
// Permission to use, copy, modify, and distribute this software in
// object code form for any purpose and without fee is hereby granted, 
// provided that the above copyright notice appears in all copies and 
// that both that copyright notice and the limited warranty and
// restricted rights notice below appear in all supporting 
// documentation.
//
// AUTODESK PROVIDES THIS PROGRAM "AS IS" AND WITH ALL FAULTS. 
// AUTODESK SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTY OF
// MERCHANTABILITY OR FITNESS FOR A PARTICULAR USE.  AUTODESK, INC. 
// DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
// UNINTERRUPTED OR ERROR FREE.
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
using System;
using System.IO;
using System.Xml;
using System.Reflection;
using System.Diagnostics;
using System.Collections;
using System.ComponentModel;
using System.Collections.Generic;
using System.Configuration.Install;
using System.Runtime.InteropServices;

using Autodesk.ADN.Utility.WinUtils;

namespace InvAddIn
{
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // AddinInstaller1: Registry Free Inventor Add-in installer
    //  
    // Author: liangx
    // Creation date: 1/24/2013 3:12:30 PM
    // 
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    [RunInstaller(true)]
    public class AddinInstaller1 : Installer
    {
        #region Base Implementation

        private System.ComponentModel.IContainer components = null;

        protected override void Dispose(bool disposing)
        {
            if (disposing && (components != null))
            {
                components.Dispose();
            }

            base.Dispose(disposing);
        }

        private void InitializeComponent()
        {
            components = new System.ComponentModel.Container();
        }

        #endregion

        public AddinInstaller1()
        {
            InitializeComponent();
        }

        public override void Install(IDictionary stateSaver)
        {
            try
            {
                base.Install(stateSaver);

                InstallUtils.InstallModeEnum installMode = InstallUtils.InstallModeEnum.kRegistryFree;

                Assembly Asm = Assembly.GetExecutingAssembly();
                FileInfo asmFile = new FileInfo(Asm.Location);

                stateSaver.Add("InstallMode", (int)installMode);

                stateSaver.Add("PathToAddinFile",
                    InstallUtils.InstallRegistryFree(
                        stateSaver,
                        Asm,
                        InstallUtils.RegFreeModeEnum.kVersionIndep,
                        string.Empty));

                // Example for version dependent

                //stateSaver.Add("PathToAddinFile",
                //    InstallUtils.InstallRegistryFree(
                //        stateSaver,
                //        Asm,
                //        InstallUtils.RegFreeModeEnum.kVersionDep,
                //        "Inventor 2012"));

            }
            catch (InstallException ex)
            {
                throw new InstallException(ex.Message);
            }
            catch
            {
                throw new InstallException("Error installing addin!");
            }
        }

        public override void Uninstall(IDictionary savedState)
        {
            try
            {
                InstallUtils.InstallModeEnum installMode =
                    (InstallUtils.InstallModeEnum)savedState["InstallMode"];

                string pathToAddinFile = (string)savedState["PathToAddinFile"];
                InstallUtils.UninstallRegistryFree(savedState, pathToAddinFile);
            }
            catch
            {

            }

            base.Uninstall(savedState);
        }
    }
}


namespace Autodesk.ADN.Utility.WinUtils
{
    public class CommonUtils
    {
        public static string GetOsVersion(bool checkSP)
        {
            OperatingSystem os = Environment.OSVersion;

            string operatingSystem = "Unknown";

            if (os.Platform == PlatformID.Win32Windows)
            {
                //This is a pre-NT version of Windows 
                switch (os.Version.Minor)
                {
                    case 0:
                        operatingSystem = "95";
                        break;
                    case 10:
                        if (os.Version.Revision.ToString() == "2222A")
                            operatingSystem = "98SE";
                        else
                            operatingSystem = "98";
                        break;
                    case 90:
                        operatingSystem = "Me";
                        break;
                    default:
                        break;
                }
            }
            else if (os.Platform == PlatformID.Win32NT)
            {
                switch (os.Version.Major)
                {
                    case 3:
                        operatingSystem = "NT 3.51";
                        break;
                    case 4:
                        operatingSystem = "NT 4.0";
                        break;
                    case 5:
                        if (os.Version.Minor == 0)
                            operatingSystem = "2000";
                        else
                            operatingSystem = "XP";
                        break;
                    case 6:
                        if (os.Version.Minor == 0)
                            operatingSystem = "Vista";
                        else
                            operatingSystem = "7";
                        break;
                    default:
                        break;
                }
            }

            //Make sure we actually got something in our OS check 
            //We don't want to just return " Service Pack 2" or " 32-bit" 
            if (operatingSystem != "")
            {
                operatingSystem = "Windows " + operatingSystem;

                if (os.ServicePack != "" && checkSP)
                    operatingSystem += " " + os.ServicePack;
            }

            return operatingSystem;
        }
    }

    public class InstallUtils
    {
        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        //Use: Enums for install options
        //
        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        public enum InstallModeEnum
        {
            kRegistryFree = 0,
            kRegistry = 1,
            kBoth = 2
        };

        public enum RegFreeModeEnum
        {
            kVersionDep,
            kVersionIndep,
            kUserOverride
        };

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        //Use: Generates path for .addin file
        //
        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        public static string GenerateAddinFileLocation(RegFreeModeEnum mode, string version)
        {
            string user = System.Environment.UserName;

            string os = Autodesk.ADN.Utility.WinUtils.CommonUtils.GetOsVersion(false);

            switch (mode)
            {
                case RegFreeModeEnum.kVersionIndep:

                    switch (os)
                    {
                        case "Windows XP":
                            return @"C:\Documents and Settings\All Users\Application Data\Autodesk\Inventor Addins\";

                        case "Windows Vista":
                        case "Windows 7":
                            return @"C:\ProgramData\Autodesk\Inventor Addins\";

                        default:
                            break;
                    }
                    break;

                case RegFreeModeEnum.kVersionDep:

                    switch (os)
                    {
                        case "Windows XP":
                            return @"C:\Documents and Settings\All Users\Application Data\Autodesk\" + version + "\\Addins\\";

                        case "Windows Vista":
                        case "Windows 7":
                            return @"C:\ProgramData\Autodesk\" + version + "\\Addins\\";

                        default:
                            break;
                    }
                    break;

                case RegFreeModeEnum.kUserOverride:
                    switch (os)
                    {
                        case "Windows XP":
                            return @"C:\Documents and Settings\" + user + @"\Application Data\Autodesk\" + version + "\\Addins\\";

                        case "Windows Vista":
                        case "Windows 7":
                            return @"C:\Users\" + user + @"\AppData\Roaming\Autodesk\" + version + "\\Addins\\";

                        default:
                            break;
                    }
                    break;

                default:
                    break;
            }

            throw new InstallException();
        }

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        //Use: Registry-Free Install / Uninstall
        //
        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        public static string InstallRegistryFree(
            IDictionary stateSaver,
            RegFreeModeEnum mode,
            string version)
        {
            return InstallRegistryFree(stateSaver, Assembly.GetExecutingAssembly(), mode, version);
        }

        public static string InstallRegistryFree(
            IDictionary stateSaver,
            Assembly Asm,
            RegFreeModeEnum mode,
            string version)
        {
            try
            {
                FileInfo asmFile = new FileInfo(Asm.Location);

                FileInfo addinFile = null;

                foreach (FileInfo fileInfo in asmFile.Directory.GetFiles())
                {
                    if (fileInfo.Extension.ToLower() == ".addin")
                    {
                        addinFile = fileInfo;
                        break;
                    }
                }

                if (addinFile == null)
                {
                    throw new InstallException();
                }

                XmlDocument xmldoc = new XmlDocument();
                xmldoc.Load(addinFile.FullName);

                XmlNode node = xmldoc.GetElementsByTagName("Assembly")[0];

                if (node == null)
                {
                    throw new InstallException();
                }

                node.InnerText = asmFile.FullName;

                string addinFilenameDest = GenerateAddinFileLocation(mode, version);

                if (!Directory.Exists(addinFilenameDest))
                    Directory.CreateDirectory(addinFilenameDest);

                addinFilenameDest += addinFile.Name;

                if (File.Exists(addinFilenameDest))
                    File.Delete(addinFilenameDest);

                // copy the addin to target folder according to OS type and all users/separate users
                xmldoc.Save(addinFilenameDest);

                addinFile.Delete();

                return addinFilenameDest;
            }
            catch
            {
                throw new InstallException("Error installing .addin file!");
            }
        }

        public static void UninstallRegistryFree(IDictionary savedState, string pathToAddinFile)
        {
            if (File.Exists(pathToAddinFile))
                File.Delete(pathToAddinFile);
        }
    }
}

File: Inventor-Training-Material-0c232f96f36e1e0bed31b07a6a6ebb901a31dd40/Module 21 - Installer/Registry_Free_AddIn_Demo_VBNet/Registry_Free_AddIn_Demo/AddinInstaller1.vb
'/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
' Copyright (c) Autodesk, Inc. All rights reserved 
' Written by Philippe Leefsma 2011 - ADN/Developer Technical Services
'
' Permission to use, copy, modify, and distribute this software in
' object code form for any purpose and without fee is hereby granted, 
' provided that the above copyright notice appears in all copies and 
' that both that copyright notice and the limited warranty and
' restricted rights notice below appear in all supporting 
' documentation.
'
' AUTODESK PROVIDES THIS PROGRAM "AS IS" AND WITH ALL FAULTS. 
' AUTODESK SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTY OF
' MERCHANTABILITY OR FITNESS FOR A PARTICULAR USE.  AUTODESK, INC. 
' DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
' UNINTERRUPTED OR ERROR FREE.
'/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
Imports System.IO
Imports System.Xml
Imports System.Reflection
Imports System.Diagnostics
Imports System.Collections
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Configuration.Install
Imports System.Runtime.InteropServices

Namespace Registry_Free_AddIn_Demo

    '/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    ' AddinInstaller1: Registry Free Inventor Add-in installer
    '  
    ' Author: liangx
    ' Creation date: 1/24/2013 2:25:14 PM
    ' 
    '/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    <RunInstaller(True)> _
    Public Class AddinInstaller1
        Inherits Installer

#Region "Base Implementation"

        Private components As System.ComponentModel.IContainer = Nothing

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing AndAlso (components IsNot Nothing) Then
                components.Dispose()
            End If

            MyBase.Dispose(disposing)
        End Sub

        Private Sub InitializeComponent()
            components = New System.ComponentModel.Container()
        End Sub

#End Region

        Public Sub New()
            InitializeComponent()
        End Sub

        Public Overrides Sub Install(ByVal stateSaver As IDictionary)
            Try
                MyBase.Install(stateSaver)

                Dim installMode As InstallUtils.InstallModeEnum = InstallUtils.InstallModeEnum.kRegistryFree

                Dim Asm As Assembly = Assembly.GetExecutingAssembly()
                Dim asmFile As New FileInfo(Asm.Location)

                stateSaver.Add("InstallMode", CInt(installMode))


                ' Example for version dependent

                'stateSaver.Add("PathToAddinFile", _
                '    InstallUtils.InstallRegistryFree( _
                '        stateSaver, _
                '        Asm, _
                '        InstallUtils.RegFreeModeEnum.kVersionDep, _
                '        "Inventor 2012"))

                stateSaver.Add("PathToAddinFile", _
                               InstallUtils.InstallRegistryFree( _
                                   stateSaver, _
                                   Asm, _
                                   InstallUtils.RegFreeModeEnum.kVersionIndep, _
                                   String.Empty))

            Catch ex As InstallException
                Throw New InstallException(ex.Message)
            Catch
                Throw New InstallException("Error installing addin!")
            End Try
        End Sub

        Public Overrides Sub Uninstall(ByVal savedState As IDictionary)
            Try
                Dim installMode As InstallUtils.InstallModeEnum =
                    CType(savedState("InstallMode"), InstallUtils.InstallModeEnum)

                Dim pathToAddinFile As String = DirectCast(savedState("PathToAddinFile"), String)
                InstallUtils.UninstallRegistryFree(savedState, pathToAddinFile)

            Catch
            End Try

            MyBase.Uninstall(savedState)
        End Sub
    End Class


    Public Class CommonUtils

        Public Shared Function GetOsVersion(ByVal checkSP As Boolean) As String
            Dim os As OperatingSystem = Environment.OSVersion

            Dim operatingSystem As String = "Unknown"

            If os.Platform = PlatformID.Win32Windows Then
                'This is a pre-NT version of Windows 
                Select Case os.Version.Minor
                    Case 0
                        operatingSystem = "95"
                        Exit Select
                    Case 10
                        If os.Version.Revision.ToString() = "2222A" Then
                            operatingSystem = "98SE"
                        Else
                            operatingSystem = "98"
                        End If
                        Exit Select
                    Case 90
                        operatingSystem = "Me"
                        Exit Select
                    Case Else
                        Exit Select
                End Select

            ElseIf os.Platform = PlatformID.Win32NT Then
                Select Case os.Version.Major
                    Case 3
                        operatingSystem = "NT 3.51"
                        Exit Select
                    Case 4
                        operatingSystem = "NT 4.0"
                        Exit Select
                    Case 5
                        If os.Version.Minor = 0 Then
                            operatingSystem = "2000"
                        Else
                            operatingSystem = "XP"
                        End If
                        Exit Select
                    Case 6
                        If os.Version.Minor = 0 Then
                            operatingSystem = "Vista"
                        Else
                            operatingSystem = "7"
                        End If
                        Exit Select
                    Case Else
                        Exit Select
                End Select
            End If

            'Make sure we actually got something in our OS check 
            'We don't want to just return " Service Pack 2" or " 32-bit" 
            If operatingSystem <> "" Then
                operatingSystem = "Windows " & operatingSystem

                If os.ServicePack <> "" AndAlso checkSP Then
                    operatingSystem += " " & os.ServicePack
                End If
            End If

            Return operatingSystem
        End Function

    End Class

    Public Class InstallUtils

        '/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        'Use: Enums for install options
        '
        '/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        Public Enum InstallModeEnum
            kRegistryFree = 0
            kRegistry = 1
            kBoth = 2
        End Enum

        Public Enum RegFreeModeEnum
            kVersionDep
            kVersionIndep
            kUserOverride
        End Enum

        '/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        'Use: Generates path for .addin file
        '
        '/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        Public Shared Function GenerateAddinFileLocation(ByVal mode As RegFreeModeEnum, ByVal version As String) As String

            Dim user As String = System.Environment.UserName

            Dim os As String = CommonUtils.GetOsVersion(False)

            Select Case mode

                Case RegFreeModeEnum.kVersionIndep

                    Select Case os
                        Case "Windows XP"
                            Return "C:\Documents and Settings\All Users\Application Data\Autodesk\Inventor Addins\"

                        Case "Windows Vista", "Windows 7"
                            Return "C:\ProgramData\Autodesk\Inventor Addins\"
                        Case Else

                            Exit Select
                    End Select
                    Exit Select

                Case RegFreeModeEnum.kVersionDep

                    Select Case os
                        Case "Windows XP"
                            Return "C:\Documents and Settings\All Users\Application Data\Autodesk\" & version & "\Addins\"

                        Case "Windows Vista", "Windows 7"
                            Return "C:\ProgramData\Autodesk\" & version & "\Addins\"
                        Case Else

                            Exit Select
                    End Select
                    Exit Select

                Case RegFreeModeEnum.kUserOverride
                    Select Case os
                        Case "Windows XP"
                            Return "C:\Documents and Settings\" & user & "\Application Data\Autodesk\" & version & "\Addins\"

                        Case "Windows Vista", "Windows 7"
                            Return "C:\Users\" & user & "\AppData\Roaming\Autodesk\" & version & "\Addins\"
                        Case Else

                            Exit Select
                    End Select
                    Exit Select
                Case Else

                    Exit Select
            End Select

            Throw New InstallException()
        End Function

        '/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        'Use: Registry-Free Install / Uninstall
        '
        '/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        Public Shared Function InstallRegistryFree(ByVal stateSaver As IDictionary, ByVal mode As RegFreeModeEnum, ByVal version As String) As String
            Return InstallRegistryFree(stateSaver, Assembly.GetExecutingAssembly(), mode, version)
        End Function

        Public Shared Function InstallRegistryFree( _
            ByVal stateSaver As IDictionary, _
            ByVal Asm As Assembly, _
            ByVal mode As RegFreeModeEnum, _
            ByVal version As String) As String

            Try
                Dim asmFile As New FileInfo(Asm.Location)

                Dim addinFile As FileInfo = Nothing

                For Each fileInfo As FileInfo In asmFile.Directory.GetFiles()
                    If fileInfo.Extension.ToLower() = ".addin" Then
                        addinFile = fileInfo
                        Exit For
                    End If
                Next

                If addinFile Is Nothing Then
                    Throw New InstallException()
                End If

                Dim xmldoc As New XmlDocument()
                xmldoc.Load(addinFile.FullName)

                Dim node As XmlNode = xmldoc.GetElementsByTagName("Assembly")(0)

                If node Is Nothing Then
                    Throw New InstallException()
                End If

                node.InnerText = asmFile.FullName

                Dim addinFilenameDest As String = GenerateAddinFileLocation(mode, version)

                If Not Directory.Exists(addinFilenameDest) Then
                    Directory.CreateDirectory(addinFilenameDest)
                End If

                addinFilenameDest += addinFile.Name

                If File.Exists(addinFilenameDest) Then
                    File.Delete(addinFilenameDest)
                End If

                ' copy the addin to target folder according to OS type and all users/separate users
                xmldoc.Save(addinFilenameDest)

                addinFile.Delete()

                Return addinFilenameDest
            Catch
                Throw New InstallException("Error installing .addin file!")
            End Try
        End Function

        Public Shared Sub UninstallRegistryFree(ByVal savedState As IDictionary, ByVal pathToAddinFile As String)
            If File.Exists(pathToAddinFile) Then
                File.Delete(pathToAddinFile)
            End If
        End Sub
    End Class

End Namespace
